datasource/repository/data.go

```diff
@@ -0,0 +1,91 @@
+//go:generate packer-sdc mapstructure-to-hcl2 -type Config,DatasourceOutput
+package repository
+
+import (
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"github.com/hashicorp/packer-plugin-sdk/hcl2helper"
+	"github.com/hashicorp/packer-plugin-sdk/template/config"
+	"github.com/zclconf/go-cty/cty"
+)
+
+type Config struct {
+	Path string `mapstructure:"path"`
+}
+
+type Datasource struct {
+	config Config
+}
+
+type DatasourceOutput struct {
+	Head     string   `mapstructure:"head"`
+	Branches []string `mapstructure:"branches"`
+	Tags     []string `mapstructure:"tags"`
+	IsClean  bool     `mapstructure:"is_clean"`
+}
+
+func (d *Datasource) ConfigSpec() hcldec.ObjectSpec {
+	return d.config.FlatMapstructure().HCL2Spec()
+}
+
+func (d *Datasource) Configure(raws ...interface{}) error {
+	err := config.Decode(&d.config, nil, raws...)
+	if err != nil {
+		return err
+	}
+	if d.config.Path == "" {
+		d.config.Path = "."
+	}
+	return nil
+}
+
+func (d *Datasource) OutputSpec() hcldec.ObjectSpec {
+	return (&DatasourceOutput{}).FlatMapstructure().HCL2Spec()
+}
+
+func (d *Datasource) Execute() (cty.Value, error) {
+	output := DatasourceOutput{}
+	emptyOutput := hcl2helper.HCL2ValueFromConfig(output, d.OutputSpec())
+
+	openOptions := &git.PlainOpenOptions{DetectDotGit: true}
+	repo, err := git.PlainOpenWithOptions(d.config.Path, openOptions)
+	if err != nil {
+		return emptyOutput, err
+	}
+	head, err := repo.Head()
+	if err != nil {
+		return emptyOutput, err
+	}
+	worktree, err := repo.Worktree()
+	if err != nil {
+		return emptyOutput, err
+	}
+	status, err := worktree.Status()
+	if err != nil {
+		return emptyOutput, err
+	}
+	branchIter, err := repo.Branches()
+	if err != nil {
+		return emptyOutput, err
+	}
+	tagIter, err := repo.Tags()
+	if err != nil {
+		return emptyOutput, err
+	}
+
+	output.Head = head.Name().Short()
+	output.IsClean = status.IsClean()
+	output.Branches = make([]string, 0)
+	_ = branchIter.ForEach(func(reference *plumbing.Reference) error {
+		output.Branches = append(output.Branches, reference.Name().Short())
+		return nil
+	})
+	output.Tags = make([]string, 0)
```

Comment:
Tags order seems be same as in `git tag -l`, so I see issue that if this is not consistent tagging:
```
$ git tag -l
rc0.1.4
v0.0.1
v0.0.2
v0.1.0
v0.1.1
v0.1.4
v0.2.0
v0.2.1
v1.3
v1.4
```

probably it will be good to have possibility to sort returned tags using some order.  Maybe can be implemented some additional parameter for sort order fields (https://git-scm.com/docs/git-for-each-ref#Documentation/git-for-each-ref.txt-worktreepath)

For example:
```
$ git tag --sort=committerdate -l
rc0.1.4
v0.1.4
v1.3
v1.4
v0.0.1
v0.0.2
v0.1.0
v0.1.1
v0.2.0
v0.2.1
```

then also can be returned last tag as first argument from list `tags`

Reply:
Sorting is an interesting idea, but I think it might be outside the scope of this plugin. This plugin is never going to be able to do everything that git can, and I don't want it to get too bloated. I think something like this is better handled by a little bit of scripting. For example, to get the latest tag and pass it to an [input variable](https://www.packer.io/guides/hcl/variables):

```bash
latestTag=$(git tag --sort=-committerdate -l | head -n 1)
packer build -var "version=$latestTag"
```

In other words, I think this is a rare use case with a simple workaround, which would be complicated to implement well. This all adds up to make it outside the scope of the plugin. If you disagree, please open an issue and I'd be happy to talk about it.